stl/inc/algorithm

```diff
@@ -5035,6 +5046,238 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
+
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_First);
+            auto _ULast  = _Get_unwrapped(_Last);
+            auto _Size   = _Size_helper(_UFirst, _ULast);
+            auto _Result = _Shift_right_impl(_STD move(_UFirst), _STD move(_ULast), _Pos_to_shift, _Size);
+            _Seek_wrapped(_First, _Result.begin());
+            auto _Final = _First;
+            _Seek_wrapped(_Final, _Result.end());
+            return {_STD move(_First), _STD move(_Final)};
+        }
+
+        template <forward_range _Rng>
+            requires permutable<iterator_t<_Rng>>
+        constexpr borrowed_subrange_t<_Rng> operator()(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) const {
+            _STL_ASSERT(_Pos_to_shift >= 0, "shift count must be non-negative (N4910 [alg.shift]/5)");
+
+            auto _Size   = _Size_helper(_Range);
+            auto _Result = _Shift_right_impl(_Ubegin(_Range), _Uend(_Range), _Pos_to_shift, _Size);
+            auto _First  = _Rewrap_iterator(_Range, _Result.begin());
+            auto _Final  = _Rewrap_iterator(_Range, _Result.end());
+            return {_STD move(_First), _STD move(_Final)};
+        }
+
+    private:
+        struct _Unsized {};
+
+        template <range _Rng>
+        _NODISCARD static constexpr auto _Size_helper(_Rng&& _Range) {
+            if constexpr (sized_range<_Rng>) {
+                return _RANGES distance(_Range);
+            } else {
+                return _Unsized{};
+            }
+        }
+
+        template <class _It, sentinel_for<_It> _Se>
+        _NODISCARD static constexpr auto _Size_helper(const _It& _First, const _Se& _Last) {
+            if constexpr (sized_sentinel_for<_Se, _It>) {
+                return _Last - _First;
+            } else {
+                return _Unsized{};
+            }
+        }
+
+        // clang-format off
+        template <input_iterator _It, weakly_incrementable _Out>
+            requires indirectly_movable<_It, _Out>
+        _NODISCARD static constexpr _Out _Move_n_helper(_It _First, iter_difference_t<_It> _Count, _Out _Result) {
+            // clang-format on
+            for (; _Count > 0; ++_First, (void) --_Count, ++_Result) {
+                *_Result = _RANGES iter_move(_First);
+            }
+
+            return _Result;
+        }
+
+        template <class _It, class _Se, class _SizeTy>
+        _NODISCARD static constexpr subrange<_It> _Shift_right_impl(
+            _It _First, _Se _Last, iter_difference_t<_It> _Pos_to_shift, const _SizeTy _Size) {
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            if (_Pos_to_shift == 0) {
+                return {_First, _RANGES next(_First, _Last)};
+            }
+
+            constexpr bool _Is_sized = !same_as<_SizeTy, _Unsized>;
+
+            if constexpr (_Is_sized) {

```

Comment:
I do not like how we have quite different implementations between `shift_left` and `shift_right`

Above we directly check whether something is sized in the call operator. This here seems more convoluted.

Reply:
`shift_right` is inherently a more complicated algorithm, if I understand it correctly. This centralizes the size handling, which seems reasonable given that the code below is so large. If you can find a way to improve this in a followup PR, I'd be interested to see it.

Reply:
Bidirectional `shift_right` is almost identical to `shift_left`. Forward `shift_right` is a completely different beast, almost a variation of `rotate` that throws away the contents of one of the subranges. I think this style - `if constexpr` dispatch between sized and unsized versions of the bounds-checking and setup, followed by the nested loops that perform the actual work - is overall more readable than the alternative of sized and unsized versions of the entire algorithm. I sometimes carry DRY to extremes and end up writing more code to avoid duplication, but that's not the case here.